src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -139,24 +139,31 @@ static int drain_aio(std::list<librados::AioCompletion*>& handles)
 }
 
 int RadosUser::list_buckets(const DoutPrefixProvider* dpp, const std::string& marker,
-			       const std::string& end_marker, uint64_t max, bool need_stats,
-			       BucketList &buckets, optional_yield y)
+                            const std::string& end_marker, bool need_stats,
+                            std::span<RGWBucketEnt> storage,
```

Comment:
So is the point of using both span<> and storage just that the resulting length of the span<> can be different?  If so, storage would only be truncated at the end, so it won't result in more allocations, and you can just pass a marker reference.  Or, vector has a distinction between size and capacity, so that could be used.  (I've never dealt with span<> before, so I'm curious)

Reply:
thanks for the review! 

[std::span](https://en.cppreference.com/w/cpp/container/span) is new in c++20, and represents a non-owning reference to a contiguous range. it's similar to `std::string_view`, except it can either be a mutable `span<T>` or an immutable `span<const T>`

i like using it in interfaces because it's very flexible; the caller can either point directly at an array on the stack, or to a preallocated vector. i used this pattern with [sal::ConfigStore](https://github.com/ceph/ceph/blob/389ea16/src/rgw/rgw_sal_config.h#L34-L41) and was happy with the result. in most cases like [get_zones_pool_set()](https://github.com/ceph/ceph/blob/389ea16/src/rgw/rgw_zone.cc#L512-L516), i was able to use a `std::array` for the storage. i ended up using `std::vector` for all of the `list_buckets()` callers just because we have a config option `rgw_list_buckets_max_chunk` that can vary

and you're right that `BucketList::buckets` is a separate span because we may return fewer entries than the requested `storage.size()`. we might instead return the number of actual entries, but i think span is nice there too because it allows the caller to loop over the results directly with `for (auto& e : listing.buckets)`